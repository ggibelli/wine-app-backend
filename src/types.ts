export enum Province {
  AG = 'AG',
  AL = 'AL',
  AN = 'AN',
  AO = 'AO',
  AR = 'AR',
  AP = 'AP',
  AT = 'AT',
  AV = 'AV',
  BA = 'BA',
  BL = 'BL',
  BN = 'BN',
  BG = 'BG',
  BI = 'BI',
  BO = 'BO',
  BZ = 'BZ',
  BS = 'BS',
  BR = 'BR',
  CA = 'CA',
  CL = 'CL',
  CB = 'CB',
  CE = 'CE',
  CT = 'CT',
  CZ = 'CZ',
  CH = 'CH',
  CO = 'CO',
  CS = 'CS',
  CR = 'CR',
  KR = 'KR',
  CN = 'CN',
  EN = 'EN',
  FM = 'FM',
  FE = 'FE',
  FI = 'FI',
  FG = 'FG',
  FO = 'FO',
  FC = 'FC',
  FR = 'FR',
  GE = 'GE',
  GO = 'GO',
  GR = 'GR',
  IM = 'IM',
  IS = 'IS',
  SP = 'SP',
  AQ = 'AQ',
  LT = 'LT',
  LE = 'LE',
  LC = 'LC',
  LI = 'LI',
  LO = 'LO',
  LU = 'LU',
  MC = 'MC',
  MN = 'MN',
  MS = 'MS',
  MT = 'MT',
  ME = 'ME',
  MI = 'MI',
  MO = 'MO',
  MB = 'MB',
  NA = 'NA',
  NO = 'NO',
  NU = 'NU',
  OR = 'OR',
  PD = 'PD',
  PA = 'PA',
  PR = 'PR',
  PV = 'PV',
  PG = 'PG',
  PS = 'PS',
  PU = 'PU',
  PE = 'PE',
  PC = 'PC',
  PI = 'PI',
  PT = 'PT',
  PN = 'PN',
  PZ = 'PZ',
  PO = 'PO',
  RG = 'RG',
  RA = 'RA',
  RC = 'RC',
  RE = 'RE',
  RI = 'RI',
  RN = 'RN',
  RM = 'RM',
  RO = 'RO',
  SA = 'SA',
  SS = 'SS',
  SV = 'SV',
  SI = 'SI',
  SR = 'SR',
  SO = 'SO',
  TA = 'TA',
  TE = 'TE',
  TR = 'TR',
  TO = 'TO',
  TP = 'TP',
  TN = 'TN',
  TV = 'TV',
  TS = 'TS',
  UD = 'UD',
  VA = 'VA',
  VE = 'VE',
  VB = 'VB',
  VC = 'VC',
  VR = 'VR',
  VV = 'VV',
  VI = 'VI',
  VT = 'VT',
}

export enum Regioni {
  ABRUZZO = 'ABRUZZO',
  BASILICATA = 'BASILICATA',
  CALABRIA = 'CALABRIA',
  CAMPANIA = 'CAMPANIA',
  EMILIA = 'EMILIA',
  FRIULI = 'FRIULI',
  LAZIO = 'LAZIO',
  LIGURIA = 'LIGURIA',
  LOMBARDIA = 'LOMBARDIA',
  MARCHE = 'MARCHE',
  MOLISE = 'MOLISE',
  PIEMONTE = 'PIEMONTE',
  PUGLIA = 'PUGLIA',
  SARDEGNA = 'SARDEGNA',
  SICILIA = 'SICILIA',
  TOSCANA = 'TOSCANA',
  BOLZANO = 'BOLZANO',
  TRENTO = 'TRENTO',
  UMBRIA = 'UMBRIA',
  VALDAOSTA = 'VALDAOSTA',
  VENETO = 'VENETO',
}

export interface Address {
  via: string;
  CAP: number;
  comune: string;
  provincia: Province;
  regione: Regioni;
}

export enum TypeAd {
  SELL = 'SELL',
  BUY = 'BUY',
}

export enum MetodoProduttivo {
  CONVENZIONALE = 'CONVENZIONALE',
  BIOLOGICO = 'BIOLOGICO',
  BIODINAMICO = 'BIODINAMICO',
  NATURALE = 'NATURALE',
}

export enum TypeProduct {
  ADWINE = 'AdWine',
  ADGRAPE = 'AdGrape',
}

export enum EspressioneComunitaria {
  DOP = 'DOP',
  IGP = 'IGP',
  ND = 'ND',
}

export enum DenomZona {
  DOC = 'DOC',
  DOCG = 'DOCG',
  IGT = 'IGT',
  VARIETALE = 'VARIETALE',
  VINO = 'VINO',
}

export enum Menzione {
  CLASSICO = 'CLASSICO',
  RISERVA = 'RISERVA',
  SUPERIORE = 'SUPERIORE',
  VIGNA = 'VIGNA',
}

export enum QueryOrderBy {
  createdAt_ASC = 'createdAt_ASC',
  createdAt_DESC = 'createdAt_DESC',
  price_ASC = 'price_ASC',
  price_DESC = 'price_DESC',
}

export interface Errors {
  text: string;
  name: string;
}

export enum Rating {
  POOR = 'POOR',
  AVERAGE = 'AVERAGE',
  OK = 'OK',
  GOOD = 'GOOD',
  PERFECT = 'PERFECT',
}

export enum Colore {
  BIANCA = 'BIANCA',
  ROSSA = 'ROSSA',
}
