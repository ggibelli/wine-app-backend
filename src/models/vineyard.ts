import mongoose, { Schema, Document } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';

import { Colore } from '../types';

export interface IVineyard {
  name: string;
  colore?: Colore;
}

export interface IVineyardDoc extends IVineyard, Document {}

export interface VineyardGraphQl extends IVineyard {
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const vineyardSchemaFields: Record<keyof IVineyard, any> = {
  name: {
    type: String,
    required: true,
  },
  colore: {
    type: String,
    enum: ['BIANCA', 'ROSSA'],
  },
};

const vineyardSchema = new Schema(vineyardSchemaFields);

vineyardSchema.plugin(mongooseUniqueValidator);

export const Vineyard = mongoose.model<IVineyardDoc>(
  'Vineyard',
  vineyardSchema
);
