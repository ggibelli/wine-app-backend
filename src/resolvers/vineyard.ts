import {
  Resolvers,
  VineyardInput,
  VineyardInputUpdate,
} from '../generated/graphql';
import Users from '../data-sources/users';
import Vineyards from '../data-sources/vineyards';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  vineyards: Vineyards;
}

const resolver: StringIndexed<Resolvers> = {
  Query: {
    async vineyards(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.vineyards.getVineyards();
    },
    async vineyard(
      _,
      { id },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.vineyards.getVineyard(id);
    },
  },
  Mutation: {
    async createVineyard(
      _,
      { vineyard }: { vineyard: VineyardInput },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.vineyards.createVineyard(vineyard);
    },
    async updateVineyard(
      _,
      { vineyard }: { vineyard: VineyardInputUpdate },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.vineyards.updateVineyard(vineyard);
    },
    async deleteVineyard(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.vineyards.deleteVineyard(id);
    },
  },
};

export default resolver;
