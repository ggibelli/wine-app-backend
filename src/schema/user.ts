import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  type ProducedWines {
    wine: Wine!
    bottlesProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  type OwnedVineyards {
    vineyard: Vineyard!
    tonsProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  input ProducedWinesInput {
    wine: ID!
    bottlesProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input OwnedVineyardsInput {
    vineyard: ID!
    tonsProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input UserInput {
    email: String!
    password: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: AddressInput!
    hideContact: Boolean!
    #verified: Boolean
    #premium: Boolean
    #ads: [ID!]
    #negotiations: [ID!]
    #reviews: [ID!]
    #adsRemaining: Int
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  input UserInputUpdate {
    _id: ID!
    email: String
    password: String
    firstName: String
    lastName: String
    pIva: String
    phoneNumber: String
    address: AddressInput
    isVerified: Boolean
    isPremium: Boolean
    hideContact: Boolean!
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  type Coordinates {
    latitude: Float
    longitude: Float
  }

  type User {
    _id: ID!
    email: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: Address!
    isVerified: Boolean!
    coordinates: Coordinates
    isPremium: Boolean
    isAdmin: Boolean!
    hideContact: Boolean!
    ads: [Ad!]
    savedAds: [Ad!]
    messages: [Message!]
    negotiations: [Negotiation!]
    reviews: [Review!]
    adsRemaining: Int @authorized
    dateCreated: Date! @date
    producedWines: ProducedWines
    ownedVineyards: OwnedVineyards
  }

  type AuthUser {
    token: String!
    user: User!
  }

  type AuthUserPayload {
    response: AuthUser
    errors: [Errors]
  }

  type UserPayload {
    response: User
    errors: [Errors]
  }

  extend type Query {
    users: [User!] @authenticated
    user(id: ID!): User @authenticated
    me: User @authenticated
  }

  extend type Mutation {
    createUser(user: UserInput!): AuthUserPayload
    updateUser(user: UserInputUpdate!): UserPayload @authenticated
    deleteUser(id: ID!): UserPayload @authenticated
    login(email: String!, password: String!): AuthUserPayload
  }
`;
