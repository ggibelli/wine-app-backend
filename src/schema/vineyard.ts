import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  type Vineyard {
    _id: ID!
    name: String!
    colore: Colore!
  }

  input VineyardInput {
    name: String!
    colore: Colore!
  }

  input VineyardInputUpdate {
    name: String
    colore: Colore
  }

  extend type Query {
    vineyards(name: String, colore: String): [Vineyard!] @authenticated
    vineyard(id: ID!): Vineyard @authenticated
  }

  extend type Mutation {
    createVineyard(vineyard: VineyardInput): Vineyard @authorized
    updateVineyard(vineyard: VineyardInputUpdate, id: ID!): Vineyard @authorized
    deleteVineyard(id: ID!): Vineyard @authorized
  }
`;
