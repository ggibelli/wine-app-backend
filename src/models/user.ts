/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
