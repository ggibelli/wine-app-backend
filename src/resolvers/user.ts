import { Resolvers, UserInput, UserInputUpdate } from '../generated/graphql';
import { UserGraphQl } from '../models/user';
import Users from '../data-sources/users';
import Ads from '../data-sources/ads';
import Messages from '../data-sources/messages';
import Negotiations from '../data-sources/negotiations';
import { ForbiddenError } from 'apollo-server-express';
import Reviews from '../data-sources/reviews';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  ads: Ads;
  messages: Messages;
  negotiations: Negotiations;
  reviews: Reviews;
  // wines: Wines;
}

export const resolver: StringIndexed<Resolvers> = {
  Query: {
    async users(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.users.getUsers();
    },
    async user(_, { id }, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.users.getUser(id);
    },
    me(_, __, { user }: { user: UserGraphQl }) {
      return user;
    },
  },
  Mutation: {
    async createUser(
      _,
      { user }: { user: UserInput },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.createUser(user);
    },
    async updateUser(
      _,
      { user }: { user: UserInputUpdate },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.updateUser(user);
    },
    async deleteUser(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.deleteUser(id);
    },
    async login(
      _,
      { email, password }: { email: string; password: string },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.login(email, password);
    },
  },
  User: {
    async ads(user, _, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.ads.getAdsByUser(user._id);
    },
    async messages(
      root,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.messages.getMessages();
      }
      throw new ForbiddenError('You can only see your own messages');
    },
    async negotiations(
      root,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.negotiations.getNegotiations();
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    async reviews(
      root,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.reviews.getReviews();
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    email(root, _, { user }: { user: UserGraphQl }) {
      if (root._id.toString() === user._id.toString()) {
        return root.email;
      }
      if (!root.hideContact) {
        return root.email;
      }
      return 'The user prefers to hide the email';
    },
    phoneNumber(root, _, { user }: { user: UserGraphQl }) {
      if (root._id.toString() === user._id.toString()) {
        return root.phoneNumber;
      }
      if (!root.hideContact) {
        return root.phoneNumber;
      }
      return 'The user prefers to hide the phone number';
    },
    // aggiungo logica per mostrare solo negotiation e certi dati solo se sono i miei
  },
};
