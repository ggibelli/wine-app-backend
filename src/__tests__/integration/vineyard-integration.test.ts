/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
//import { ObjectId } from 'mongodb';
import { User } from '../../models/user';
import { users } from '../../tests/mocksTests';
import { VineyardInput, VineyardInputUpdate } from '../../generated/graphql';
import { Colore } from '../../types';
import createVineyardDb from '../../utils/vineyardExtractor';
import { Vineyard, VineyardGraphQl } from '../../models/vineyard';

const { query, mutate, setOptions } = testClient;

const VINEYARDS = gql`
  {
    vineyards {
      name
      colore
    }
  }
`;

const VINEYARD = gql`
  query GetVineyard($id: ID!) {
    vineyard(id: $id) {
      name
      colore
    }
  }
`;

const CREATE_VINEYARD = gql`
  mutation CreateVineyard($vineyard: VineyardInput!) {
    createVineyard(vineyard: $vineyard) {
      response {
        name
        colore
      }
      errors {
        name
        text
      }
    }
  }
`;

const UPDATE_VINEYARD = gql`
  mutation UpdateVineyard($vineyard: VineyardInputUpdate!) {
    updateVineyard(vineyard: $vineyard) {
      response {
        name
        colore
      }
      errors {
        name
        text
      }
    }
  }
`;

const DELETE_VINEYARD = gql`
  mutation DeleteVineyard($id: ID!) {
    deleteVineyard(id: $id) {
      response {
        name
        colore
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const vineyardDb = createVineyardDb();
  const user = new User(usersMock[0]);
  const otherUser = new User(usersMock[1]);
  const vineyard = new Vineyard(vineyardDb[0]);
  await user.save();
  await otherUser.save();
  await vineyard.save();
});

describe('Integration test vineyards', () => {
  it('query vineyards fails if not logged in', async () => {
    const res = await query(VINEYARDS);
    expect(res).toMatchSnapshot();
  });

  it('query single vineyard fails if not logged', async () => {
    const vineyards: VineyardGraphQl[] = await Vineyard.find({}).lean().exec();
    const res = await query(VINEYARD, {
      variables: { id: vineyards[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create vineyard mutation fails if not logged in', async () => {
    const vineyard: VineyardInput = {
      name: 'Uva vinosa',
      colore: Colore.BIANCA,
    };
    const res = await mutate(CREATE_VINEYARD, {
      variables: { vineyard },
    });
    expect(res).toMatchSnapshot();
  });

  it('query vineyards succeds if logged in', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(VINEYARDS);
    expect(res).toMatchSnapshot();
  });

  it('query single vineyard succeds if logged', async () => {
    const vineyard: VineyardGraphQl[] = await Vineyard.find({}).lean().exec();
    const res = await query(VINEYARD, {
      variables: { id: vineyard[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create vineyard mutation successfull if logged in and admin', async () => {
    const vineyard: VineyardInput = {
      name: 'Uva vinosa',
      colore: Colore.BIANCA,
    };
    const res = await mutate(CREATE_VINEYARD, {
      variables: { vineyard },
    });
    expect(res).toMatchSnapshot();
  });

  it('update vineyard mutation succeds if logged in and admin', async () => {
    const vineyardToUpdate: VineyardGraphQl[] = await Vineyard.find({})
      .lean()
      .exec();
    const vineyard: VineyardInputUpdate = {
      _id: vineyardToUpdate[0]._id.toString(),
      name: 'Vinotto',
    };
    const res = await mutate(UPDATE_VINEYARD, {
      variables: { vineyard },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete vineyard mutation succeds if logged in and admin', async () => {
    const vineyardToDelete: VineyardGraphQl[] = await Vineyard.find({})
      .lean()
      .exec();
    const res = await mutate(DELETE_VINEYARD, {
      variables: { id: vineyardToDelete[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('update vineyard mutation fails if logged in and not admin', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const vineyardToUpdate: VineyardGraphQl[] = await Vineyard.find({})
      .lean()
      .exec();
    const vineyard: VineyardInputUpdate = {
      _id: vineyardToUpdate[0]._id.toString(),
      name: 'Caprotto',
    };
    const res = await mutate(UPDATE_VINEYARD, {
      variables: { vineyard },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete vineyard mutation fails if logged in and not admin', async () => {
    const vineyardToDelete: VineyardGraphQl[] = await Vineyard.find({})
      .lean()
      .exec();
    const res = await mutate(DELETE_VINEYARD, {
      variables: { id: vineyardToDelete[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create vineyard mutation fails if logged in and not admin', async () => {
    const vineyard: VineyardInput = {
      name: 'Uva vinosa',
      colore: Colore.BIANCA,
    };
    const res = await mutate(CREATE_VINEYARD, {
      variables: { vineyard },
    });
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
