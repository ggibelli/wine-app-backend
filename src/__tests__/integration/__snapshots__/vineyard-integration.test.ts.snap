// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test vineyards create vineyard mutation fails if logged in and not admin 1`] = `
Object {
  "data": Object {
    "createVineyard": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to be Admin to continue",
      "path": Array [
        "createVineyard",
      ],
    },
  ],
}
`;

exports[`Integration test vineyards create vineyard mutation fails if not logged in 1`] = `
Object {
  "data": Object {
    "createVineyard": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "createVineyard",
      ],
    },
  ],
}
`;

exports[`Integration test vineyards create vineyard mutation successfull if logged in and admin 1`] = `
Object {
  "data": Object {
    "createVineyard": Object {
      "errors": Array [],
      "response": Object {
        "colore": "BIANCA",
        "name": "Uva vinosa",
      },
    },
  },
}
`;

exports[`Integration test vineyards delete vineyard mutation fails if logged in and not admin 1`] = `
Object {
  "data": Object {
    "deleteVineyard": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to be Admin to continue",
      "path": Array [
        "deleteVineyard",
      ],
    },
  ],
}
`;

exports[`Integration test vineyards delete vineyard mutation succeds if logged in and admin 1`] = `
Object {
  "data": Object {
    "deleteVineyard": Object {
      "errors": Array [],
      "response": Object {
        "colore": null,
        "name": "Vinotto",
      },
    },
  },
}
`;

exports[`Integration test vineyards query single vineyard fails if not logged 1`] = `
Object {
  "data": Object {
    "vineyard": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "vineyard",
      ],
    },
  ],
}
`;

exports[`Integration test vineyards query single vineyard succeds if logged 1`] = `
Object {
  "data": Object {
    "vineyard": Object {
      "colore": null,
      "name": "Abrusco",
    },
  },
}
`;

exports[`Integration test vineyards query vineyards fails if not logged in 1`] = `
Object {
  "data": Object {
    "vineyards": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "vineyards",
      ],
    },
  ],
}
`;

exports[`Integration test vineyards query vineyards succeds if logged in 1`] = `
Object {
  "data": Object {
    "vineyards": Array [
      Object {
        "colore": null,
        "name": "Abrusco",
      },
    ],
  },
}
`;

exports[`Integration test vineyards update vineyard mutation fails if logged in and not admin 1`] = `
Object {
  "data": Object {
    "updateVineyard": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to be Admin to continue",
      "path": Array [
        "updateVineyard",
      ],
    },
  ],
}
`;

exports[`Integration test vineyards update vineyard mutation succeds if logged in and admin 1`] = `
Object {
  "data": Object {
    "updateVineyard": Object {
      "errors": Array [],
      "response": Object {
        "colore": null,
        "name": "Vinotto",
      },
    },
  },
}
`;
